psutil/_psibmi.py

```diff
@@ -0,0 +1,647 @@
+    """Return system per-CPU times as a list of named tuples"""
+    ret = cext.per_cpu_times()
+    return [scputimes(*x) for x in ret]
+
+
+def cpu_count_logical():
+    # if hasattr(cext, 'cpu_count_online'):
+    #     return cext.cpu_count_online()
+    """Return the number of logical CPUs in the system."""
+    cursor = _conn.cursor()
+    cursor.execute("""select CURRENT_CPU_CAPACITY
+                      from table (QSYS2.SYSTEM_STATUS()) x""")
+    ncpus = math.ceil(cursor.fetchone()[0])
+    cursor.close()
+    return ncpus
+
+
+def cpu_count_physical():
+    if hasattr(cext, 'cpu_count'):
+        return cext.cpu_count()
+    cursor = _conn.cursor()
+    cursor.execute("""select CONFIGURED_CPUS
+                      from table (QSYS2.SYSTEM_STATUS()) x""")
+    ncpus = cursor.fetchone()[0]
+    cursor.close()
+    return ncpus
+
+
+def cpu_stats():
+    """Return various CPU stats as a named tuple."""
+    return _common.scpustats(0, 0, 0, 0)  # TODO
+
+# =====================================================================
+# --- disks
+# =====================================================================
+
+
+def disk_usage(path='/'):
+    cursor = _conn.cursor()
+    cursor.execute("""SELECT SUM(UNIT_STORAGE_CAPACITY) as TOTAL,
+                             SUM(UNIT_SPACE_AVAILABLE) as AVAIL
+                             FROM QSYS2.SYSDISKSTAT""")
+    for row in cursor:
+        total = row[0]
+        avail = row[1]
+        used = total - avail
+        percent = (float(100.00*used)/float(total))
+        cursor.close()
+        return _common.sdiskusage(total, used, avail, percent)
+
+
+def disk_io_counters(all=False):  # TODO
+    cursor = _conn.cursor()
+    cursor.execute("SELECT UNIT_NUMBER FROM QSYS2.SYSDISKSTAT")
+    counters = {}
+    for row in cursor:
+        # TODO: query system performance data to get this
+        counters["unit:"+str(row[0])] = (0, 0, 0, 0, 0, 0)
+    cursor.close()
+    return counters
+
+
+def disk_partitions(all=False):
+    cursor = _conn.cursor()
+    cursor.execute("""SELECT UNIT_NUMBER,ASP_NUMBER,UNIT_TYPE,DISK_TYPE
+                      FROM QSYS2.SYSDISKSTAT""")
+    ret = []
+    for row in cursor:
+        if(row[2] == 0):
+            unit_type = "not_solid_state"
+        else:
+            unit_type = "solid_state"
+        ret.append(_common.sdiskpart(
+                   "unit:"+str(row[0]), "/", unit_type, "asp:"+str(row[1])))
+    cursor.close()
+    return ret
+
+
+# =====================================================================
+# --- network
+# =====================================================================
+
+
+net_if_addrs = cext_posix.net_if_addrs
+
+
+def net_io_counters(pernic=True):
+    cursor = _conn.cursor()
+    cursor.execute("SELECT LINE_DESCRIPTION FROM QSYS2.NETSTAT_INTERFACE_INFO")
+    counters = {}
+    for row in cursor:
+        counters[row[0]] = (0, 0, 0, 0, 0, 0, 0, 0)  # TODO
+    cursor.close()
+    return counters
```

Comment:
For this cursor stuff I suggest to use a context manager. Something like this:

```python

import contextlib


@contextlib.contextmanager
def cursor_ctx():
    cursor = _conn.cursor()
    try:
        yield cursor
    finally:
        cursor.close()
```


...and when you have to use it:

```python
with cursor_ctx() as cursor:
    cursor.execute('...')
```